src/pgzero/joystick.py

```diff
@@ -0,0 +1,577 @@
+    CR = 6
+
+
+class PS4Axis(IntEnum):
+    NUM = 6
+    LX = 0
+    LY = 1
+    LT = 4
+    RX = 2
+    RY = 3
+    RT = 5
+
+
+class PS5Btns(IntEnum):
+    NUM = 17
+    FU = 3
+    FD = 0
+    FL = 2
+    FR = 1
+    DU = 13  # Simulated D-Pad like with the Xbox controller.
+    DD = 14
+    DL = 15
+    DR = 16
+    SL = 4
+    SR = 5
+    PL = 11
+    PR = 12
+    CL = 8
+    CM = 10
+    CR = 9
+
+
+class PS5Axis(IntEnum):
+    NUM = 6
+    LX = 0
+    LY = 1
+    LT = 2
+    RX = 3
+    RY = 4
+    RT = 5
+
+
+BTN_NAMES = ("face_up", "face_down", "face_left", "face_right", "dpad_up",
+             "dpad_down", "dpad_left", "dpad_right", "shoulder_left",
+             "shoulder_right", "push_left", "push_right", "center_left",
+             "center_middle", "center_right")
+PressedBtns = namedtuple("PressedBtns", BTN_NAMES)
+
+
+class Joystick:
+    """PGZero wrapper class for joystick use."""
+
+    def __init__(self, stick):
+        self._stick = stick
+        name = self._stick.get_name()
+        # We use the system name of the controller to determine
+        # the correct mappings to use.
+        if "Series" in name:
+            self._type = "XSeries"
+            self._btn_map = XSeriesBtns
+            self._axis_map = XSeriesAxis
+        elif "Xbox" in name:
+            self._type = "Xbox"
+            self._btn_map = X360Btns
+            self._axis_map = X360Axis
+        elif "PS4" in name:
+            self._type = "PS4"
+            self._btn_map = PS4Btns
+            self._axis_map = PS4Axis
+        elif "Sony Interactive" in name:
+            self._type = "PS5"
+            self._btn_map = PS5Btns
+            self._btn_map = PS5Axis
+        elif "Switch" in name:
+            self._type = "Switch"
+            self._btn_map = SwitchBtns
+            self._axis_map = SwitchAxis
+        else:
+            self._type = "UNKNOWN"
+            # If the type is unknown, we fall back to Xbox
+            # mappings since these are most common among
+            # generic controller layouts.
+            self._btn_map = X360Btns
+            self._axis_map = X360Axis
+
+        self._pressed = [False] * self._btn_map.NUM
+        self._axis = [0] * self._axis_map.NUM
+        # Since triggers don't start centered but rather unpressed,
+        # their starting values are -1.
+        self._axis[self._axis_map.RT] = -1
+        self._axis[self._axis_map.LT] = -1
+
+        # These variables are used to give a lockout time for physical
+        # HAT button simulation to prevent ghost inputs.
+        self._DU_open = True
+        self._DD_open = True
+        self._DL_open = True
+        self._DR_open = True
+
+    # Used with clock.schedule to prevent ghost button presses via timeouts.
+    def _unlock_DU(self):
+        self._DU_open = True
+
+    def _unlock_DD(self):
+        self._DD_open = True
+
+    def _unlock_DL(self):
+        self._DL_open = True
+
+    def _unlock_DR(self):
+        self._DR_open = True
```

Comment:
I can see the point of this debouncing but I think the entire thing should be encapsulated separately from this class and probably using game ticks instead of the clock class. The clock is likely to be pausable and support slowmo (based on work that is already in Wasabi2D).

Reply:
Actually if we do that maybe we just make a "master clock" for pgzero internal use, so that would be Ok